### PR TITLE
fix: explicitly begin failed implicit begin transactions

### LINKED ISSUE #4516: atomicity is violated if the first operation in a RW Transaction fails
Given the following example (pseudocode, we should write an actual test)

```cc
1:  Transaction txn = MakeReadWriteTransaction();
2:  auto result1 = client_->ExecuteDml(txn, "INSERT that will fail (duplicate row)");
3:  auto result2 = client_->ExecuteDml(txn, "INSERT that will succeed");
4:  auto commit_result = client_->Commit(txn, {});
```

In our client library, although the first `ExecuteDml` fails, the second succeeds, as does the `Commit`.
This behavior may be seen as violating the atomicity guarantee - users may expect that the failure of the first `ExecuteDml` means that the transaction cannot successfully Commit.  From http://goto.google.com/spanner-txns#atomicity-consistency-durability

>In addition to the Isolation property, Spanner provides Atomicity (all writes must succeed in order for the transaction to commit; if one or more writes within the transaction fail, the entire transaction fails), [...]

The reason our client library behaves this way is because we defer `Transaction` begin until the first operation that uses the transaction as an optimization. If the first operation on a `Transaction` is a write (DML) that fails, the Spanner backend does not actually begin the transaction or return one in the response, the C++ `Transaction` object remains in its initial state; assuming all subsequent DML operations succeed, the `Commit` will also succeed.

### PR DESCRIPTION
**Pull request by mr-salty**
If an operation which has `TransactionSelector::begin` set in order to
implicitly begin a transaction fails, explicitly call `BeginTransaction`.

If that fails, invalidate the `Transaction` so all subsequent operations
fail - most importantly `Commit`, which is necessary to maintain atomicity
since the first operation did not execute.

Otherwise, retry the original operation using the transaction ID
obtained from `BeginTransaction`. That may fail or succeed, but even if
it fails, it may be a "non-constraint error" (read errors and certain
types of write/dml errors) which permit the rest of the `Transaction` to
succeed. Beginning the transaction leaves it to the backend to make
that decision.

Fixes #4516

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4706)

### CODE REVIEW

**Comment by mr-salty** [comment]
let me take a look at the build failures... it passed all unit tests and integration tests before I committed.

**Comment by mr-salty** [comment]
PTAL - forgot to check ok() in one place before dereferencing (and the first test failures I looked at sent me down the wrong rathole, it looked like some gmock issue)... should have looked at asan first - and also #4707 which might have made this more obvious.

**Comment by mr-salty** [comment]
well, apparently the gmock thing was an actual issue. It sometimes thinks one of the `GrpcReader` calls is a sequence prerequisite for one of the `SpannerStub` calls, I don't understand why (is there some threading involved here?  even that doesn't seem to fit the symptoms) but I may be holding `InSequence` incorrectly, especially combined with a lambda. I think this more explicit version should work but I'm not able to reproduce the problem reliably.

In any case, I think it's an issue with the test and not the code so I think you can still review it. If the latest commit doesn't fix it I'll take another look in the morning.

**Comment by mr-salty** [comment]
looks like that fixed the test issues although Brad and I were discussing a change to handing the case where we get an "OK" response and no transaction - we should probably handle that differently... although I think that could be done in a follow-up PR, it's a corner case that probably won't occur in practice (and the checked in code does not handle it so it's not a regression).

File: google/cloud/spanner/internal/connection_impl.cc [hunk 1]
```diff
@@ -436,33 +440,47 @@ StatusOr<std::vector<ReadPartition>> ConnectionImpl::PartitionReadImpl(
...
+                spanner_proto::PartitionReadRequest const& request) {
+          return stub->PartitionRead(context, request);
+        },
+        request, __func__);
+    if (s->has_begin()) {
+      if (response.ok() && response->has_transaction()) {
+        s->set_id(response->transaction().id());
+      } else {
+        // A transaction should have been returned but was not.
+        auto begin = BeginTransaction(session, s->begin(), __func__);
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.cc:457]: I don't think we should do this if the response was OK.

File: google/cloud/spanner/internal/connection_impl.cc [hunk 2]
```diff
@@ -436,33 +440,47 @@ StatusOr<std::vector<ReadPartition>> ConnectionImpl::PartitionReadImpl(
...
+        if (!begin.ok()) {
+          s = begin.status();  // invalidate the transaction
+          return begin.status();
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.cc:460]: And, given that the response was not OK, I think this should return `response.status()`.

File: google/cloud/spanner/internal/connection_impl.cc [hunk 3]
```diff
@@ -436,33 +440,47 @@ StatusOr<std::vector<ReadPartition>> ConnectionImpl::PartitionReadImpl(
...
+        }
+        s->set_id(begin->id());
+        *request.mutable_transaction() = *s;
+        continue;
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.cc:464]: In general, I thought the control flow would look like:
```
{
  auto request = <setup-request>();
  for (;;) {
    *request.mutable_transaction() = *s;
    auto response = <handle-request>(request);
    if (s->has_begin()) {
      if (!response) {
        auto begin = BeginTransaction(...);
        if (!begin) {
          s = begin.status();
          return response;
        }
        s->set_id(begin->id);
        continue;  // try again with id
      }
      if (response->has_transaction()) {
        s->set_id(response->transaction().id());
      }
    }
    return <handle-response>(response);
  } 
} 
```

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 4]
```diff
@@ -402,6 +409,51 @@ TEST(ConnectionImplTest, ReadImplicitBeginTransaction) {
   EXPECT_THAT(txn, HasSessionAndTransactionId("test-session-name", "ABCDEF00"));
 }
 
+TEST(ConnectionImplTest, ReadImplicitBeginTransactionPermanentFailure) {
+  auto mock = std::make_shared<spanner_testing::MockSpannerStub>();
+
+  auto db = Database("dummy_project", "dummy_instance", "dummy_database_id");
+  auto conn = MakeLimitedRetryConnection(db, mock);
+
+  auto make_reader = []() {
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc:418]: Nit: I believe we choose to omit the `()` when there are no parameters.

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 5]
```diff
@@ -402,6 +409,51 @@ TEST(ConnectionImplTest, ReadImplicitBeginTransaction) {
...
+    grpc::Status grpc_status(grpc::StatusCode::PERMISSION_DENIED, "uh-oh");
+    auto reader = absl::make_unique<MockGrpcReader>();
+    EXPECT_CALL(*reader, Read(_)).WillOnce(Return(false));
+    EXPECT_CALL(*reader, Finish()).WillOnce(Return(grpc_status));
+    return reader;
+  };
+  // TODO(salty) some interaction between the spanner calls and the reader
+  // calls seems to confuse gmock and cause some flakiness. Does explicitly
+  // sequencing them fix the issue?
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc:427]: Is this resolved?

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 6]
```diff
@@ -402,6 +409,51 @@ TEST(ConnectionImplTest, ReadImplicitBeginTransaction) {
...
+      .WillOnce(Return(MakeTestTransaction()));
+  EXPECT_CALL(*mock, StreamingRead(_, _))
+      .InSequence(s)
+      .WillOnce(Return(ByMove(std::move(reader2))));
+
+  Transaction txn = MakeReadOnlyTransaction(Transaction::ReadOnlyOptions());
+  auto rows = conn->Read({txn, "table", KeySet::All(), {"UserId", "UserName"}});
+  for (auto& row : rows) {
+    EXPECT_EQ(StatusCode::kPermissionDenied, row.status().code());
+    EXPECT_THAT(row.status().message(), HasSubstr("uh-oh"));
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc:453]: Use `StatusIs()` matcher?

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 7]
```diff
@@ -677,6 +748,37 @@ TEST(ConnectionImplTest, ExecuteDmlDeleteTooManyTransientFailures) {
   EXPECT_THAT(result.status().message(), HasSubstr("try-again in ExecuteDml"));
 }
 
+// Tests that a Transaction that fails to begin does not successfully commit -
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc:751]: s/ -/ as/?  Or, two sentences?

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 8]
```diff
@@ -677,6 +748,37 @@ TEST(ConnectionImplTest, ExecuteDmlDeleteTooManyTransientFailures) {
...
+  Status status(StatusCode::kInvalidArgument, "cannot begin transaction");
+  {
+    InSequence seq;
+    EXPECT_CALL(*mock, BatchCreateSessions(_, _))
+        .WillOnce(Return(MakeSessionsResponse({"session-name"})));
+
+    // The first `ExecuteDml` call tries to implicitly begin the transaction
+    // via `ExecuteSql`, and then explicitly via `BeginTransaction`. Both fail,
+    // and we should receive no further RPC calls - since the transaction is
+    // not valid ther client fails any subsequent operations itself.
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc:768]: s/ther/the/

**Comment by mr-salty** [comment]
___
*[google/cloud/spanner/internal/connection_impl.cc, line 457 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4706#-MDIxMy6_7MWigBSuS4q-r2-457:-MDMJY_jDWpf-FilHOWl:b-apxsvx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4da49406b20a6420ea7fa8dfb6385dfb6e67fed6/google/cloud/spanner/internal/connection_impl.cc#L457)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I don't think we should do this if the response was OK.
</blockquote></details>

my earlier comment: Brad and I were discussing a change to handing the case where we get an "OK" response and no transaction - we should probably handle that differently... although I think that could be done in a follow-up PR, it's a corner case that probably won't occur in practice (and the checked in code does not handle it so it's not a regression).

**Comment by mr-salty** [comment]
___
*[google/cloud/spanner/internal/connection_impl_test.cc, line 453 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4706#-MDIxMy7wgVIf6XymFEF-r2-453:-MDNtKkh1dxN8jQdtkJV:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4da49406b20a6420ea7fa8dfb6385dfb6e67fed6/google/cloud/spanner/internal/connection_impl_test.cc#L453)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looked like a new use.
</blockquote></details>

Done

File: google/cloud/spanner/internal/connection_impl.cc [hunk 9]
```diff
@@ -389,19 +398,22 @@ RowStream ConnectionImpl::ReadImpl(
...
-      auto metadata = reader.ok() ? (*reader)->Metadata() : absl::nullopt;
-      if (metadata && metadata->has_transaction()) {
-        s->set_id(metadata->transaction().id());
+      if (reader.ok()) {
+        auto metadata = (*reader)->Metadata();
+        if (metadata && metadata->has_transaction()) {
+          s->set_id(metadata->transaction().id());
+        } else {
+          s = MissingTransactionStatus(__func__);
+          return MakeStatusOnlyResult<RowStream>(s.status());
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.cc:407]: Observation: It still seems strange to me to fail an operation that actually succeeded (despite the missing transaction ID), but I guess I'm happy in the knowledge that it never happens.

**Pull request merged**